pallets/circuit/src/machine/mod.rs

```diff
@@ -0,0 +1,577 @@
+                        let updated_fsx = Bids::<T>::try_bid(
+                            &mut current_fsx,
+                            bid_amount,
+                            &bidder,
+                            &requester,
+                            sfx_id,
+                            local_ctx.xtx_id,
+                        )?;
+
+                        Self::update_current_step_fsx(local_ctx, &updated_fsx);
+
+                        Some(CircuitStatus::InBidding)
+                    },
+                    _ => return Err(Error::<T>::BiddingInactive),
+                }
+            },
+            PrecompileResult::Continue => None,
+            // Assume kill attempt with fallible post_update to be intended as infallible cleanup to kill op
+            //  in case fallible post_update passes, proceed with kill op
+            // ToDo: check between allowed status enforcements - kill status / allowed enforced status
+            PrecompileResult::TryKill(cause) => Some(CircuitStatus::Killed(cause)),
+            PrecompileResult::ForceUpdateStatus(force_status) => {
+                if CircuitStatus::InBidding == status && force_status == CircuitStatus::Ready {
+                    SquareUp::<T>::bind_bidders(local_ctx);
+                }
+                Some(force_status)
+            },
+            PrecompileResult::Revert(cause) => Some(CircuitStatus::Reverted(cause)),
+        };
+        let status_change = Self::update_status(local_ctx, enforced_new_status)?;
+        post_update(status_change.clone(), local_ctx)?;
+        Ok(Self::apply(local_ctx, status_change))
+    }
+
+    pub fn load_xtx(xtx_id: XtxId<T>) -> Result<LocalXtxCtx<T>, Error<T>> {
+        let xtx = <pallet::Pallet<T> as Store>::XExecSignals::get(xtx_id)
+            .ok_or(Error::<T>::XtxDoesNotExist)?;
+        let full_side_effects = <pallet::Pallet<T> as Store>::FullSideEffects::get(xtx_id)
+            .ok_or(Error::<T>::SetupFailedXtxStorageArtifactsNotFound)?;
+        let local_state = <pallet::Pallet<T> as Store>::LocalXtxStates::get(xtx_id)
+            .ok_or(Error::<T>::SetupFailedXtxStorageArtifactsNotFound)?;
+
+        Ok(LocalXtxCtx {
+            local_state,
+            use_protocol: UniversalSideEffectsProtocol::new(),
+            xtx_id,
+            xtx,
+            full_side_effects,
+        })
+    }
+
+    fn update_current_step_fsx(
+        local_ctx: &mut LocalXtxCtx<T>,
+        updated_fsx: &Vec<
+            FullSideEffect<
+                <T as frame_system::Config>::AccountId,
+                <T as frame_system::Config>::BlockNumber,
+                BalanceOf<T>,
+            >,
+        >,
+    ) {
+        let (current_step, _) = local_ctx.xtx.steps_cnt;
+
+        match local_ctx.full_side_effects.get_mut(current_step as usize) {
+            Some(current_step) => {
+                *current_step = updated_fsx.to_vec();
```

Comment:
Not crazy about this one, as we're doing a lot of writes the we probably don't need to do. Wonder how this affects the state transition proofs the collators build.

Reply:
the state in `apply`will update all of the FSX anyways. could expose more specific `PrecompileResult` for updating entries on FSX one by one - TryBid is already there, will do the same for TryConfirm